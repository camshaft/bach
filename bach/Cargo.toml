[package]
name = "bach"
version = "0.0.7"
authors = ["Cameron Bytheway <bytheway.cameron@gmail.com>"]
description = "Discrete-event simulation environment for async workflows"
repository = "https://github.com/camshaft/bach"
license = "MIT"
edition = "2021"

[features]
coop = []
full = ["coop", "metrics", "net", "tracing"]
metrics = ["dep:metrics"]
net = []
tracing = ["dep:tracing"]

[dependencies]
arr_macro = "0.2"
async-task = { version = "4", default-features = false }
atomic-waker = "1"
bolero-generator.workspace = true
event-listener-strategy = { version = "0.5.2", default-features = false }
futures-core = { version = "0.3", default-features = false }
intrusive-collections = "0.9"
pin-project-lite = "0.2"
metrics = { version = "0.24", optional = true }
rand = { version = "0.8", default-features = false }
rand_xoshiro = "0.6"
tracing = { version = "0.1", optional = true }

[dev-dependencies]
bolero.workspace = true

[lints.rust.unexpected_cfgs]
level = "warn"
check-cfg = [
    'cfg(always_disabled)', # used to always disable something
]
