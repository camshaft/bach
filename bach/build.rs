fn main() {
    let ac = autocfg::new();
    ac.emit_path_cfg("core::task::Waker::data", "feature_waker_data");
}
