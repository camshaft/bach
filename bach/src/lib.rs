extern crate alloc;

#[macro_use]
mod tracing;
#[macro_use]
pub mod metrics;

pub mod coop;
pub mod environment;
pub mod executor;
pub mod ext;
pub mod group;
#[cfg(any(test, feature = "net"))]
pub mod net;
pub mod rand;
pub mod scope;
pub mod sync;
pub mod task;
pub mod time;

/// Returns `true` if the caller is being executed in a `bach` environment
pub fn is_active() -> bool {
    task::scope::try_borrow_with(|scope| scope.is_some())
}
