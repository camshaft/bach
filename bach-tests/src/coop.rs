use bach::{environment::default::Runtime, ext::*, sync::queue::vec_deque::Queue};
use std::{collections::HashMap, sync::Mutex};

fn sim(f: impl Fn()) -> impl Fn() {
    crate::testing::init_tracing();
    move || {
        let mut rt = Runtime::new().with_coop(true).with_rand(None);
        rt.run(&f);
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
#[allow(dead_code)]
enum Event {
    Start,
    Message {
        receiver: u8,
        sender_group: u8,
        sender_id: u8,
    },
    ReceiverClose {
        receiver: u8,
    },
}

impl Event {
    fn check(v: &[Event]) -> &[Event] {
        let mut group = vec![];
        let mut seen = HashMap::<_, usize>::new();
        for event in v {
            match *event {
                Event::Start => {
                    let group = std::mem::take(&mut group);
                    *seen.entry(group).or_default() += 1;
                }
                _ => {
                    group.push(*event);
                }
            }
        }
        *seen.entry(group).or_default() += 1;

        let mut duplicate = false;

        for (group, count) in seen {
            if count == 1 {
                continue;
            }
            duplicate = true;
            eprintln!("duplicate ({count}): {group:#?}");
        }

        assert!(!duplicate, "duplicate interleavings found");

        v
    }
}

#[test]
fn interleavings() {
    static LOG: Mutex<Vec<Event>> = Mutex::new(vec![]);

    bolero::check!().exhaustive().run(sim(|| {
        LOG.lock().unwrap().push(Event::Start);

        for group in 0..2 {
            let (sender, receiver) = Queue::builder().with_capacity(Some(20)).build().channel();

            async move {
                while let Ok((sender_group, sender_id)) = receiver.pop().await {
                    LOG.lock().unwrap().push(Event::Message {
                        receiver: group,
                        sender_group,
                        sender_id,
                    });
                }

                LOG.lock()
                    .unwrap()
                    .push(Event::ReceiverClose { receiver: group });
            }
            .primary()
            .spawn_named(format!("[{group}] server"));

            for id in 0..2 {
                let sender = sender.clone();
                async move {
                    for _ in 0..1 {
                        sender.push((group, id)).await.unwrap();
                    }
                }
                .primary()
                .spawn_named(format!("[{group}] client{id}"));
            }
        }
    }));

    insta::assert_debug_snapshot!(Event::check(&*LOG.lock().unwrap()));
}

#[test]
fn joined_interleavings() {
    static LOG: Mutex<Vec<Event>> = Mutex::new(vec![]);

    bolero::check!().exhaustive().run(sim(|| {
        LOG.lock().unwrap().push(Event::Start);
        eprintln!("start");

        let (sender, receiver) = Queue::builder().with_capacity(Some(20)).build().channel();

        for group in 0..2 {
            let receiver = receiver.clone();
            async move {
                while let Ok((sender_group, sender_id)) = receiver.pop().await {
                    LOG.lock().unwrap().push(Event::Message {
                        receiver: group,
                        sender_group,
                        sender_id,
                    });
                }

                LOG.lock()
                    .unwrap()
                    .push(Event::ReceiverClose { receiver: group });
            }
            .primary()
            .spawn_named(format!("[{group}] server"));

            for id in 0..1 {
                let sender = sender.clone();
                async move {
                    for _ in 0..1 {
                        sender.push((group, id)).await.unwrap();
                    }
                }
                .primary()
                .spawn_named(format!("[{group}] client{id}"));
            }
        }
    }));

    insta::assert_debug_snapshot!(Event::check(&*LOG.lock().unwrap()));
}
