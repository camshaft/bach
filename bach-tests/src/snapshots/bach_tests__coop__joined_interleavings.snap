---
source: bach-tests/src/coop.rs
expression: "Event::check(&*LOG.lock().unwrap())"
---
[
    Start,
    Message {
        receiver: 0,
        sender_group: 0,
        sender_id: 0,
    },
    Message {
        receiver: 1,
        sender_group: 1,
        sender_id: 0,
    },
    ReceiverClose {
        receiver: 0,
    },
    ReceiverClose {
        receiver: 1,
    },
    Start,
    Message {
        receiver: 0,
        sender_group: 0,
        sender_id: 0,
    },
    Message {
        receiver: 1,
        sender_group: 1,
        sender_id: 0,
    },
    ReceiverClose {
        receiver: 1,
    },
    ReceiverClose {
        receiver: 0,
    },
    Start,
    Message {
        receiver: 1,
        sender_group: 1,
        sender_id: 0,
    },
    Message {
        receiver: 0,
        sender_group: 0,
        sender_id: 0,
    },
    ReceiverClose {
        receiver: 1,
    },
    ReceiverClose {
        receiver: 0,
    },
    Start,
    Message {
        receiver: 1,
        sender_group: 1,
        sender_id: 0,
    },
    Message {
        receiver: 0,
        sender_group: 0,
        sender_id: 0,
    },
    ReceiverClose {
        receiver: 0,
    },
    ReceiverClose {
        receiver: 1,
    },
]
